opensoundscape/annotations.py

```diff
@@ -914,3 +850,46 @@ def one_hot_to_categorical(one_hot, classes):
     """
     classes = np.array(classes)
     return [list(classes[np.array(row).astype(bool)]) for row in one_hot]
+
+
+def find_overlapping_idxs_in_clip_df(
+    annotation_start,
+    annotation_end,
+    clip_df,
+    min_label_overlap,
+    min_label_fraction=None,
+):
+    """
+    Finds the (file, start_time, end_time) index values for the rows in the clip_df that overlap with the annotation_start and annotation_end
+    Args:
+        annotation_start: start time of the annotation
+        annotation_end: end time of the annotation
+        clip_df: dataframe with multi-index ['file', 'start_time', 'end_time']
+        min_label_overlap: minimum duration (seconds) of annotation within the
+    Returns:
+        [(file, start_time, end_time)]) Multi-index values for the rows in the clip_df that overlap with the annotation_start and annotation_end.
+    """
+    # ignore all rows that start after the annotation ends. Start is level 1 of multi-index
+    clip_df = clip_df.loc[clip_df.index.get_level_values(1) < annotation_end]
+    # and all rows that end before the annotation starts. End is level 2 of multi-index
+    clip_df = clip_df.loc[clip_df.index.get_level_values(2) > annotation_start]
+    # now for each time-window, calculate the overlaps
+    clip_df["overlap"] = [
+        overlap([annotation_start, annotation_end], [row[1], row[2]])
+        for row in clip_df.index
+    ]
+
+    # discard annotations that do not overlap with the time window
+    clip_df = clip_df[clip_df["overlap"] > 0]
+
+    # calculate the fraction of each annotation that overlaps with this time window
+    clip_df["overlap_fraction"] = [
+        overlap_fraction([annotation_start, annotation_end], [row[1], row[2]])
+        for row in clip_df.index
+    ]
+
+    if min_label_overlap is not None:
+        clip_df = clip_df[clip_df["overlap"] >= min_label_overlap]
+    if min_label_fraction is not None:
```

Comment:
this does not reproduce the desired behavior of "min_label_fraction". The desired behavior may be a bit confusing, and the argument could probably have a better name. See the old implementation and docstring of `one_hot_labels_on_time_interval`: the idea is that if you specify min_label_fraction=0.5, the annotation will be retained if >=0.5 of the annotation overlaps with a clip, _even if_ the absolute time duration of the overlap is < min_label_overlap. This is useful for short annotations in particular: imagine min_label_overlap=0.25 but some annotations are 0.2 seconds long - they are always discarded and don't appear on any clips! So setting min_label_overlap=1.0 (or 0.5 to catch labels on the boundaries of clips) avoids tossing out labels entirely. 

As implemented, short clips are removed from clip_df regardless of the min_label_fraction argument.

A verbose but more descriptive name for the argument would be `keep_labels_overlaping_clip_by_this_ratio_or_more_even_if_short`

Reply:
I guess we need a test for that behavior also, since no tests failed in this implementation

Reply:
Thank you for catching this. I understand now:
- min_label_overlap is a required argument, and will always be passed.
- min_label_fraction is an optional argument. If passed, it ensures annotations with at least min_label_fraction are included, even if they have less than min_label_overlap seconds of overlap.

I wrote a test for this behavior.